Pure async RpcClient

devhawk opened the pull request:
This PR plumbs removes all sync RpcClient calls and plumbs async thru end to end. 

If someone wants to block on an RpcClient call, they can use `.Result` (or `.GetAwaiter().GetResult()`) on the top level Task object.

~~Note, there are a few tests that are failing, so this PR isn't quite ready to merge yet. But I wanted to open it for review feedback since the core design is complete.~~ All the tests are passing now.

fixes #209

ProDog commented:
Need wait https://github.com/neo-project/neo-modules/issues/311 fixed. 
Otherwise NetworkFee is inaccurate.

devhawk commented:
> Need wait #311 fixed. Otherwise NetworkFee is inaccurate.

I'm fine with waiting, but RpcClient is currently building against Neo 3.0.0-preview3-00. I'm thinking it would make more sense to merge this PR as is and then update RpcClient to a more recent build of Neo to get the correct NetworkFee as a separate task.

shargon reviewed (commented):
It's not recommended to add `Async` at the end of the method name?

shargon commented on src/RpcClient/TransactionManager.cs:
```diff
@@ -212,14 +177,88 @@ public TransactionManager Sign()
+        /// Calculate NetworkFee
+        /// </summary>
+        /// <returns></returns>
+        private async Task<long> CalculateNetworkFee()
```
We should wait for https://github.com/neo-project/neo-modules/pull/338

devhawk commented:
> It's not recommended to add `Async` at the end of the method name?

I didn't think that it was when you don't provide sync and async versions of methods. However, based on [official MSFT docs](https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/task-based-asynchronous-pattern-tap), it appears that I should: " Asynchronous methods in TAP include the Async suffix after the operation name for methods that return awaitable types, such as Task, Task<TResult>, ValueTask, and ValueTask<TResult>." Will fix

devhawk commented:
Now that #338 is merged and #311 is fixed, I've updated this PR and we're ready to merge.

Note, I've added the fix for #343 to this PR. `TransactionManager` now supports an optional magic constructor parameter (defaults to ProtocolSettings.Default.Magic if unspecified) which is then passed to a recently added overload of `Neo.Wallets.Helper.Sign`

shargon commented:
I will review it

vncoelho commented on src/RpcClient/ContractClient.cs:
```diff
@@ -30,10 +32,10 @@ public ContractClient(RpcClient rpc)
         /// <param name="args">operation arguments</param>
         /// <returns></returns>
-        public RpcInvokeResult TestInvoke(UInt160 scriptHash, string operation, params object[] args)
+        public Task<RpcInvokeResult> TestInvokeAsync(UInt160 scriptHash, string operation, params object[] args)
```
This one is not `public async`?

devhawk replied:
no, because it doesn't use await. Since TestInvoke doesn't need to process the results of InvokeScriptAsync, it can simply return the Task<RpcInvokeResult> returned by InvokeScriptAsync

devhawk commented:
@shargon, are we still waiting on anythign for this PR?

shargon commented:
> @shargon, are we still waiting on anything for this PR?

No, we can review and merge now

shargon commented on src/RpcClient/TransactionManager.cs:
```diff
@@ -40,12 +49,10 @@ private class SignItem { public Contract Contract; public HashSet<KeyPair> KeyPa
         /// TransactionManager Constructor
         /// </summary>
         /// <param name="rpc">the RPC client to call NEO RPC API</param>
-        /// <param name="sender">the account script hash of sender</param>
```
Add summary for `magic` ?

shargon commented on src/RpcClient/TransactionManager.cs:
```diff
@@ -56,22 +63,28 @@ public TransactionManager(RpcClient rpc)
             context = new ContractParametersContext(Tx);
-            signStore = new List<SignItem>();
+
+            QueueWork(async _ =>
```
This will not guarantee that the transaction will be with `ValidUntilBlock` and `SystemFee` set at the end of the method, will it? because the Tx it's public and can be used before `Sign`

devhawk replied:
This is true. I was trying to maintain the fluent API, but if this is a concern then MakeTransaction needs to be async. will fix

shargon commented on src/RpcClient/TransactionManager.cs:
```diff
@@ -162,39 +161,80 @@ public TransactionManager AddWitness(UInt160 scriptHash, params object[] paramet
+        public async Task<Transaction> SignAsync()
         {
+            // wait for all queued work to complete
+            await fluentOperationsTask;
```
if we call again `MakeTransaction` the queue will be empty?

devhawk replied:
If you call `MakeTransaction` again, it will queue two new items.

Based on this + the earlier comment about using the Tx before `ValidUntilBlock` and `SystemFee` are set, I'm going to redesign this API a bit.

shargon reviewed (commented):
@cloud8little could you test these changes?

devhawk commented:
Note to self - need to add magic parameter to helper methods that create TxManager (CreateDeployContractTxAsync, CreateTransferTxAsync, CreateTransferTxAsync)

devhawk commented:
Based on @shargon's observation that Tx is publicly accessible before `ValidUntilBlock` and `SystemFee`, I've redesgined the api so that MakeTransaction is async. I've moved MakeTransactionAsync to a new class TransactionManagerFactory to make it clear that a TransactionManager instance is only good for signing a single transaction. This approach allowed me to simplify TransactionManager and remove the QueueWork infrastructure. It's not as fluid as before - there are two async operations instead of one - but it ensures the incomplete Tx instance is never visible.

Usage:

``` csharp
var factory = new TransactionManagerFactory(rpcClient, magic);
var manager = await factory.MakeTransactionAsync(script, signers);
var tx = await manager
    .AddSignature(key)
    .SignAsync();
```

I've also provided a helper for scenarios when the user is only creating a single `TransactionManager`:

``` csharp
var manager = await TransactionManager.MakeTransactionAsync(rpcClient, script, signers, attributes, magic);
var tx = await manager
    .AddSignature(key)
    .SignAsync();
```

shargon commented:
@superboyiii can your team test it?

Tommo-L commented:
Waiting for some days, they are busy testing the DeFi projects.

cloud8little commented on src/RpcClient/RpcClient.cs:
```diff
@@ -335,33 +342,36 @@ public UInt256 SubmitBlock(byte[] block)
+        public async Task<RpcInvokeResult> InvokeScriptAsync(byte[] script, params Signer[] signers)
         {
             List<JObject> parameters = new List<JObject> { script.ToHexString() };
             if (signers.Length > 0)
```
invokescript also need sender as input. 
https://github.com/neo-project/neo-modules/blob/fed3f7a87d65d91bbe4e42bce94734cc76571d5c/src/RpcServer/RpcServer.SmartContract.cs#L117-L123

Tommo-L replied:
see https://github.com/neo-project/neo-modules/pull/368

cloud8little commented on src/RpcClient/ContractClient.cs:
```diff
@@ -54,13 +63,11 @@ public Transaction CreateDeployContractTx(byte[] contractScript, ContractManifes
-            Transaction tx = new TransactionManager(rpcClient)
-                .MakeTransaction(script, signers)
+            TransactionManagerFactory factory = new TransactionManagerFactory(rpcClient, magic);
+            TransactionManager manager = await factory.MakeTransactionAsync(script, signers).ConfigureAwait(false);
```
need sender as input.

devhawk replied:
Didn't realize RpcServer had changed w/o updating RpcClient. Will fix

cloud8little commented on src/RpcClient/Nep5API.cs:
```diff
@@ -105,19 +113,18 @@ public RpcNep5TokenInfo GetTokenInfo(UInt160 scriptHash)
 
-            return tx;
+            TransactionManagerFactory factory = new TransactionManagerFactory(rpcClient, magic);
+            TransactionManager manager = await factory.MakeTransactionAsync(script, signers).ConfigureAwait(false);
```
need sender as input because invokescript rpc request need.

cloud8little commented on src/RpcClient/Nep5API.cs:
```diff
@@ -130,21 +137,20 @@ public Transaction CreateTransferTx(UInt160 scriptHash, KeyPair fromKey, UInt160
 
-            return tx;
+            TransactionManagerFactory factory = new TransactionManagerFactory(rpcClient, magic);
+            TransactionManager manager = await factory.MakeTransactionAsync(script, signers).ConfigureAwait(false);
```
need sender also.

cloud8little commented:
@devhawk use the previous commit f3b06f5 and https://github.com/neo-project/neo-modules/pull/368, it works well for sendrawtransaction. I will continue testing nep5 methods and basic rpc commands.

devhawk commented:
@cloud8little I reset my topic branch to f3b06f5c72fc7a21430ecd37f8feed96869843a5 and force pushed

cloud8little commented:
All tests passed, including nep5APIs name/symbol/totalSupply and getblock*, getrawtransaction/sendrawtransaction; send*; transactionmanager, dump/import private key,etc.

devhawk commented on src/RpcClient/RpcClient.cs:
```diff
@@ -196,83 +192,89 @@ public static ContractState ContractStateFromJson(JObject json)
         /// Returns the current NEO consensus nodes information and voting status.
         /// </summary>
-        public RpcValidator[] GetValidators()
+        public async Task<RpcValidator[]> GetValidatorsAsync()
```
Can do, but it would be nice to get this PR merged and then make updates to master than keep having to adjust this PR to other changes. I'm going on vacation later this week, so I won't be able to keep adapting this PR in advance of preview 4

shargon commented on src/RpcClient/WalletAPI.cs:
```diff
@@ -113,12 +114,12 @@ public Transaction ClaimGas(string key)
-            BigInteger balance = nep5API.BalanceOf(NativeContract.NEO.Hash, toHash);
-            Transaction transaction = nep5API.CreateTransferTx(NativeContract.NEO.Hash, keyPair, toHash, balance);
-            rpcClient.SendRawTransaction(transaction);
+            BigInteger balance = await nep5API.BalanceOfAsync(NativeContract.NEO.Hash, toHash).ConfigureAwait(false);
```
Maybe we can save the balance of, and transfer 0 tokens

shargon replied:
https://github.com/neo-project/neo/blob/6886991ca67f2c2148429ba87bf7d782f8cb12c8/src/neo/SmartContract/Native/Tokens/NeoToken.cs#L49

devhawk replied:
I think we should limit this PR to focus on making the existing RpcClient async. Suggestions like this can be made in later PRs and after we ship preview 4.

superboyiii commented:
@erikzhang Shall you review as well? I will merge it one day later if no oppose since many PRs is waiting for this to be merged.
